### PR TITLE
feat: Cluster terminal enhancement I2

### PR DESCRIPTION
**Pull request by milindk8**
# Description

Include cluster terminal FE-only fixes & enhancement

1. Cluster terminal: Auto-select a node on which the terminal can be launched
2. Show node-groups -  Add node groups column on node list Ability to search nodes by node groups
                                      - Terminal > Node dropdown
                                      - Categorize nodes by node groups
                                      - Categorize ungrouped nodes in ‘Independent nodes’ category
                                      - If no node group exists, show all nodes without any categorization
3. For custom images select sh as the default
4. Ability to provide custom shell
5. Messaging if the pod has been deleted
6. When switching shell > Show message eg. Switching shell to bash
7.  If the selected shell is not supported, show the error message “{shellName} is not supported in the selected image”
8. Connection timed out handling

Fixes #([AB1893](https://dev.azure.com/DevtronLabs/Devtron/_boards/board/t/CSAT/Stories/?workitem=1893))

Screenshot 2023-02-15 at 10 57 54 AM

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] By selecting the auto node option to select a node which connects automatically
- [x] With the BE help testing error cases for node connection

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/common/helpers/Helpers.tsx [hunk 1]
```diff
@@ -1207,3 +1208,23 @@ export const k8sStyledAgeToSeconds = (duration: string): number => {
...
+    }, {})
+
+    const groupList = Object.entries(objList).map(([key, value]) => {
+        return {
+            label: key,
+            options: value,
+        }
+    })
+
+    return [{label: '',options: [AUTO_SELECT]},...groupList]
```
Comment (vivek-devtron) [inline src/components/common/helpers/Helpers.tsx:1229]: Format this file

File: src/components/ClusterNodes/NodeList.tsx [hunk 2]
```diff
@@ -295,7 +296,19 @@ export default function NodeList({ imageList, isSuperAdmin, namespaceList }: Clu
                 if (searchedTextMap.size !== matchedLabelCount) {
                     continue
                 }
+            } else if (selectedSearchTextType === 'nodeGroup') {
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/NodeList.tsx:299]: we should create some constant for nodeGroup as it is being used on multiple places

File: src/components/ClusterNodes/ClusterEvents.tsx [hunk 3]
```diff
@@ -2,11 +2,18 @@ import React, { useState, useEffect } from 'react'
...
 import MessageUI from '../v2/common/message.ui'
 import { getClusterEvents } from './clusterNodes.service'
 
-export default function ClusterEvents({ terminalAccessId }: { terminalAccessId: number }) {
-    const [events, setEvents] = useState([])
+export default function ClusterEvents({
+    terminalAccessId,
+    reconnectStart,
+}: {
+    terminalAccessId: number
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/ClusterEvents.tsx:13]: Please move this to type

File: src/components/ClusterNodes/ClusterEvents.tsx [hunk 4]
```diff
@@ -15,12 +22,13 @@ export default function ClusterEvents({ terminalAccessId }: { terminalAccessId:
             getClusterEvents(terminalAccessId)
                 .then((response) => {
                     setLoading(false)
-                    const events = response.result?.events.items
+                    const events = response.result
                     setEvents(events)
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/ClusterEvents.tsx:26]: we can directly set the data into setEvents no need to create the variable. also please accomodatae `events?.eventsResponse?.events.items` it here only and and not at the time of passing the props on line no 45

File: src/components/ClusterNodes/ClusterTerminal.tsx [hunk 5]
```diff
@@ -148,7 +168,10 @@ export default function ClusterTerminal({
                         setSocketConnection(SocketConnectionType.DISCONNECTED)
                     })
             }
-            toggleOptionChange()
+
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/ClusterTerminal.tsx:171]: This is supposed to be in useEffect's clean up function. right? Please move it out from try catch block

File: src/components/ClusterNodes/ClusterTerminal.tsx [hunk 6]
```diff
@@ -436,26 +476,49 @@ export default function ClusterTerminal({
...
                             <div style={{ minWidth: '145px' }}>
                                 <ReactSelect
                                     placeholder="Select Containers"
-                                    options={clusterNodeList}
+                                    options={nodeGroups}
                                     defaultValue={selectedNodeName}
                                     value={selectedNodeName}
                                     onChange={onChangeNodes}
-                                    styles={clusterSelectStyle}
+                                    styles={{
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/ClusterTerminal.tsx:487]: move this out to const

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 7]
```diff
@@ -188,42 +198,52 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
         }
     }
 
-    const preFetchData = (status = '', firstMessageReceived = false) => {
+    const preFetchData = (podState = '', status = '') => {  
         const _terminal = terminal
-
+        let startingText = 'create'
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:203]: we should create constant for this

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 8]
```diff
@@ -188,42 +198,52 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
...
-            _terminal.write('Connecting to pod terminal.')
-        } else if (status === CLUSTER_STATUS.FAILED) {
-            _terminal.write(' \u001b[38;5;196mFailed\u001b[0m')
-            _terminal.write(' | \u001b[38;5;110m\u001b[4mCheck Pod Events\u001b[0m')
-            _terminal.write(' | ')
-            _terminal.write('\u001b[38;5;110m\u001b[4mCheck Pod Manifest\u001b[0m')
-        } else {
-            _terminal.write('..')
+        if(terminalViewProps.isShellSwitched){
+            startingText = 'shell'
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:209]: same here

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 9]
```diff
@@ -244,7 +266,6 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
             action: `Selected Pod`,
             label: `${terminalViewProps.nodeName}/${terminalViewProps.containerName}/${terminalViewProps.shell.value}`,
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:267]: we should create some common method to getLabel and use that on multiple places

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 10]
```diff
@@ -452,36 +479,71 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
         setIsReconnection(true)
     }
 
+    const switchTOPodEventTab = () => {
+        terminalViewProps.setTerminalTab(1)
+    }
+
+    const renderErrorMessageStrip = (errorMessage) => {
+        if (errorMessage.message === 'timedOut') {
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:487]: we should move this to const

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 11]
```diff
@@ -452,36 +479,71 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
...
+                    <u
+                        className="cursor"
+                        onClick={onClickResume}
+                    >
+                        {TERMINAL_TEXT.RETRY_CONNECTION}
+                    </u>&nbsp;
+                    {TERMINAL_TEXT.CASE_OF_ERROR}
+                </div>
+            )
+        } else if (errorMessage.message === 'Terminated') {
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:504]: same here

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 12]
```diff
@@ -369,18 +384,30 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
             .then((response: any) => {
                 let sessionId = response.result.userTerminalSessionId
                 let status = response.result.status
-                if (!sessionId && count) {
+                if(status === 'Running' && !response.result?.isValidShell){
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:387]: we should move statuses to const

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 13]
```diff
@@ -369,18 +384,30 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
...
+                    preFetchData(status,'failed')
+                    setErrorMessage({message: response.result?.errorReason, reason: ''})
+                } else if (status === 'Terminated'){
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:390]: same here

File: src/components/ClusterNodes/ClusterTerminal.tsx (hunk 5 above)
Comment (milindk8) [reply src/components/ClusterNodes/ClusterTerminal.tsx:172]: no its not

File: src/components/v2/common/ReactSelect.utils.tsx [hunk 14]
```diff
@@ -201,10 +201,11 @@ export const noMatchingPlatformOptions = (): string => {
 }
 
 export function GroupHeading(props) {
-    if (!props.data.label) return null
+    const {data, hideClusterName} = props
+    if (!data.label) return null
     return (
         <components.GroupHeading {...props}>
-            <div className="flex dc__no-text-transform flex-justify h-100">Cluster : {props.data.label}</div>
+            <div className="flex dc__no-text-transform flex-justify h-100">{hideClusterName ? '' : 'Cluster : '} {data.label}</div>
```
Comment (shivani170) [inline src/components/v2/common/ReactSelect.utils.tsx:208]: flip the condition

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/constants.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    AUTO_SELECT_NODE: 'autoSelectNode',
+    SHELL: 'shell',
+    TIMEDOUT: 'timedOut',
+    FAILED: 'failed',
+    SUCCEDED: 'Succeded',
+    TERMINATED: 'Terminated',
+    RUNNING: 'Running', 
+}
+
+export const GA_CONST = {
```
Comment (shivani170) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/constants.ts:25]: Can we rename it more specifically somewhat like `TERMINAL_RESOURCE_GA`

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/EventsTable.tsx [hunk 16]
```diff
@@ -19,9 +20,18 @@ export function EventsTable({ loading, eventsList, isResourceBrowserView }: Even
             if (eventsList && eventsList.length > 0) {
                 return (
                     <div className="cn-0 ">
+                        {errorValue.status === 'Terminated' && <div className="pl-20 h-24 flex left pr-20 w-100 bcr-7 cn-0">
```
Comment (shivani170) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/EventsTable.tsx:23]: Please extract `Terminated` from coonstant

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 17]
```diff
@@ -28,20 +29,22 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
     const [firstMessageReceived, setFirstMessageReceived] = useState(false)
     const [popupText, setPopupText] = useState<boolean>(false)
     const isOnline = useOnline()
-    const [errorMessage, setErrorMessage] = useState<string>('')
+    const [errorMessage, setErrorMessage] = useState({message: '', reason: ''})
```
Comment (shivani170) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:32]: Please add type

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 18]
```diff
@@ -28,20 +29,22 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
...
     const socketConnectionRef = useRef<SocketConnectionType>(terminalViewProps.socketConnection)
     const { serverMode } = useContext(mainContext)
-
+    const autoSelectNodeRef = useRef('')
+    const prevNode = useRef('')
```
Comment (shivani170) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:36]: can we rename it to  `prevNodeRef` &  `currNodeRef`

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx [hunk 19]
```diff
@@ -138,7 +141,7 @@ function TerminalView(terminalViewProps: TerminalViewProps) {
 
         _socket.onopen = function () {
             if (terminalViewProps.isClusterTerminal) {
-                preFetchData(CLUSTER_STATUS.RUNNING, true)
+                preFetchData(CLUSTER_STATUS.RUNNING, 'Succeded')
```
Comment (shivani170) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/Terminal.tsx:144]: `Succeded` extract out in constant

File: src/components/ClusterNodes/clusterNodes.service.tsx [hunk 20]
```diff
@@ -49,12 +49,12 @@ export const updateNodeManifest = (
     return put(`${Routes.NODE_CAPACITY}?clusterId=${clusterId}&name=${nodeName}`, nodeData)
 }
 
-export const clusterTerminalStart = (data): Promise<ResponseType> => {
-    return post(`${Routes.CLUSTER_TERMINAL}/${Routes.START}`, data)
+export const clusterTerminalStart = (data, option): Promise<ResponseType> => {
```
Comment (shivani170) [inline src/components/ClusterNodes/clusterNodes.service.tsx:52]: Please add type for data and option

File: src/components/ClusterNodes/NodeListSearchFilter.tsx [hunk 21]
```diff
@@ -149,6 +156,7 @@ export default function NodeListSearchFliter({
                                 {[
                                     { value: 1, label: 'name', type: 'main' },
                                     { value: 2, label: 'label', type: 'main' },
+                                    { value: 3, label: 'nodeGroup', type: 'main' },
```
Comment (shivani170) [inline src/components/ClusterNodes/NodeListSearchFilter.tsx:159]: Please keep name,label & node group in constant

File: src/components/ClusterNodes/NodeList.tsx [hunk 22]
```diff
@@ -11,6 +11,8 @@ import {
     filterImageList,
     showError,
     useBreadcrumb,
+    createClusterEnvGroup,
```
Comment (shivani170) [inline src/components/ClusterNodes/NodeList.tsx:14]: Please remove this

**Review by shivani170 — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**